File: 41-first-missing-positive/first-missing-positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int i = 0; 

        while (i < nums.size()) {
            if (nums[i] > 0){
                int pos = nums[i] - 1;
                if (pos <= nums.size() - 1 && nums[pos] != nums[i]) swap(nums[pos], nums[i]);
                else i ++;
            } else i ++;
            
        }

        for(i = 0; i < nums.size(); i++) {
            if (nums[i] != i+1) break;
        }

        return i + 1;
    }
};

File: 41-first-missing-positive/first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        i = 0

        while i < len(nums):
            if nums[i] <= 0: i += 1
            else:
                pos = nums[i] - 1
                if pos < len(nums) and nums[i] != nums[pos]:
                    nums[i], nums[pos] = nums[pos], nums[i]
                else: i += 1
        
        i = 0

        while i < len(nums):
            if i+1 != nums[i]: return i + 1
            i += 1
        
        return i + 1